### PR TITLE
controller: add TriggeredBy custom watches to engine

### PR DESCRIPTION
**Pull request by sttts**
### Description of your changes

Added `controller.TriggeredBy` to create controller engine watches based on a different informer cache (the one from outside).

This is helpful e.g. to [trigger the compositor reconciling XRs through changes on `CompositionRevision`](https://github.com/crossplane/crossplane/pull/4582). The compositor has its own informer cache because it is dynamically started for each XR kind, done using the controller engine. This kind of controllers though today can only watch objects in the same cache. This PR adds a way to inject watches from another cache, e.g. the top-level cache watching the `CompositionRevision` kind.

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested

This has been tested as part of https://github.com/crossplane/crossplane/pull/4582, manually for the demo in that PR, plus I am working on tests for it.

[contribution process]: https://git.io/fj2m9

### CODE REVIEW

**Comment by negz** [comment]
@sttts This LGTM, but can you update the PR description with some context as to why we want this. Any context on how it's been tested would also help.

**Comment by sttts** [comment]
Added context.

File: pkg/controller/engine.go
Comment (ytsarev) [inline pkg/controller/engine.go:167]: I assume it was tested e2e by https://github.com/crossplane/crossplane/pull/4582/files#diff-52e214a3e1aa3741fcbc3dd3faca4e6ae54e98b177690300591a4d147294a70dR416 ? 

Do we want to extend unit test at https://github.com/crossplane/crossplane-runtime/blob/master/pkg/controller/engine_test.go ?

Comment (sttts) [reply pkg/controller/engine.go:167]: The engine is basically untested today. It only checks some errors.

Comment (ytsarev) [reply pkg/controller/engine.go:167]: ok, so maybe we should separately track extending unit test coverage? I'll leave it up to you and not hold this one :)

**Review by ytsarev — APPROVED** [review]
LGTM

File: pkg/controller/engine.go
Comment (sttts) [reply pkg/controller/engine.go:167]: I tried now for 20min. You mock yourself to death with cache, informer, controller. At the end, you are not sure anymore what you are testing. There is a reason this is only lightly tested 🤷

Comment (ytsarev) [reply pkg/controller/engine.go:167]: @sttts got it, thanks a lot for checking! 👍

Comment (negz) [reply pkg/controller/engine.go:167]: Yeah, this whole package is kinda just syntactic sugar. 😬

**Review by negz — APPROVED** [review]

**Comment by negz** [comment]
Thanks @sttts! The use case makes sense to me now.

**Comment by sttts** [comment]
What is missing to merge this?

File: pkg/controller/engine.go
Comment (sttts) [inline pkg/controller/engine.go:170]: @negz see that I slightly changed the signature. It's more flexible now.

Before: `TriggeredBy(someCache, kind, h, p)`
Now: `TriggeredBy(source.Kind(someCache, kind), h, p)`

**Pull request merged**
